### PR TITLE
Fix/navigation/top app bar UI tests

### PR DESCRIPTION
**Pull request by agonzalez-r**
# Fix UI tests for M1: TopAppBar

## Description
This PR closes issue #90. It completes all features of the `TopAppBar` by adding an edit button! 
As for the back button, this feature is conditional, and can be set up by passing as argument `editButton = True` to the `TopAppBar()` composable, along with the a function of the desired navigation action when this button is clicked (passed as the `onEditButtonClick` argument). 

I must remark that no further tests were added, the tests for pre-existing features of the top bar were already very extensive and complete. Navigation should not be tested in `TopAppBarTest`. 

#### Added
none
#### Modified
- `TopAppBar.kt`: created arguments `editButton` (boolean, to indicate if this button should exist for the respective screen that calls the function) and `onEditButtonClick` (nullable lambda function, where the navigation action upon click should be specified)

#### Removed
none

## Testing
- `TopAppBarTest.kt`: added tests for `editButton` feature by ensuring its presence when argument respective argument is true, that upon click the `onEditButtonClick` function is called, that exception is thrown if no such function is provided in this case, and other edge cases...

## Screenshots
Screenshot 2024-10-29 at 22 18 53

### CODE REVIEW

**Comment by francelu** [comment]
@agonzalez-r  
Please assign yourself :) PO

File: app/src/main/java/com/android/periodpals/ui/navigation/TopAppBar.kt [hunk 1]
```diff
@@ -45,14 +46,26 @@ import com.android.periodpals.ui.theme.PurpleGrey80
...
  */
 @OptIn(ExperimentalMaterial3Api::class)
 @Composable
-fun TopAppBar(title: String, backButton: Boolean = false, onBackButtonClick: (() -> Unit)? = {}) {
+fun TopAppBar(
+    title: String,
+    backButton: Boolean = false,
+    editButton: Boolean = false,
+    onBackButtonClick: (() -> Unit)? = {},
+    onEditButtonClick: (() -> Unit)? = {}
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/navigation/TopAppBar.kt:60]: I suggest setting `onBackButtonClick` and `onEditButtonClick` to null by default instead of assigning them empty actions (`{}`). This way, we avoid showing buttons without actions, ensuring buttons only appear when functional. For example:
```
backButton: Boolean = false,
onBackButtonClick: (() -> Unit)? = null,
editButton: Boolean = false,
onEditButtonClick: (() -> Unit)? = null
```

This approach keeps the button display tied directly to its action, making the component’s behaviour more intuitive.

File: app/src/androidTest/java/com/android/periodpals/ui/navigation/TopAppBarTest.kt [hunk 2]
```diff
@@ -29,6 +29,14 @@ class TopAppBarTest {
...
   }
 
+  @Test
+  fun topAppBar_displaysEditButton() {
+    composeTestRule.setContent { TopAppBar(title = "Tampon Timer", editButton = true) }
+
+    composeTestRule.onNodeWithTag("topBar").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("editButton").assertIsDisplayed()
+  }
+
```
Comment (francelu) [inline app/src/androidTest/java/com/android/periodpals/ui/navigation/TopAppBarTest.kt:39]: After updating `TopAppBar`, add `onEditButtonClick = {}` and verify the back button is hidden with `.assertDoesNotExist()`.  
Apply the same check in `topAppBar_displaysBackButton()`.

File: app/src/androidTest/java/com/android/periodpals/ui/navigation/TopAppBarTest.kt [hunk 3]
```diff
@@ -44,6 +52,21 @@ class TopAppBarTest {
...
+    var editButtonClicked = false
+
+    composeTestRule.setContent {
+      TopAppBar(
+          title = "Tampon Timer",
+          editButton = true,
+          onEditButtonClick = { editButtonClicked = true })
+    }
+
+    composeTestRule.onNodeWithTag("editButton").performClick()
```
Comment (francelu) [inline app/src/androidTest/java/com/android/periodpals/ui/navigation/TopAppBarTest.kt:66]: It would be nice to check here too that go back button isn't displayed. Same for `topAppBar_BackButtonClick()`.  
If you want even more consistency, you could check for `topAppBar_displaysTitle()` that both buttons aren't displayed.

File: app/src/androidTest/java/com/android/periodpals/ui/navigation/TopAppBarTest.kt [hunk 4]
```diff
@@ -72,6 +114,15 @@ class TopAppBarTest {
...
 
+  @Test
+  fun topAppBar_editButtonTrue_onEditButtonClickNotNull_doesNotThrowException() {
+    composeTestRule.setContent {
+      TopAppBar(title = "Test Title", editButton = true, onEditButtonClick = { /* Do nothing */})
+    }
+    composeTestRule.onNodeWithTag("topBar").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("editButton").assertIsDisplayed()
+  }
+
```
Comment (francelu) [inline app/src/androidTest/java/com/android/periodpals/ui/navigation/TopAppBarTest.kt:125]: This test is irrelevant after doing the changes I mentioned earlier. This would be similar to `topAppBar_displaysEditButton()`.  
Same goes for `topAppBar_backButtonTrue_onBackButtonClickNotNull_doesNotThrowException()`.

**Review by francelu — CHANGES_REQUESTED** [review]
## Summary
You did a great job implementing the edit button! However, the tests need a few changes to improve quality, primarily by setting the on-click actions to `null` instead of `{}`. More details are in the code comments.

## Important
#### Code Quality
- Improve navigation by using `null` for default on-click actions.

#### Functionality
- Ensure buttons are displayed only when actions are assigned.

## Testing
- Update tests to check for button visibility based on action assignments.

## Documentation
- Ensure any changes in parameters are reflected in the documentation.

## Suggestions
- **Review and implement the proposed changes in the code comments.**

## Steps before PR approved
1. Change the default values for `onBackButtonClick` and `onEditButtonClick` to `null`.
2. Update tests to verify that buttons do not display when actions are `null`.

**Comment by francelu** [comment]
I realize there are quite a few changes in the tests, mainly because I didn't test thoroughly while implementing navigation. Sorry for the extra work!

File: app/src/main/java/com/android/periodpals/ui/navigation/TopAppBar.kt (hunk 1 above)
Comment (agonzalez-r) [reply app/src/main/java/com/android/periodpals/ui/navigation/TopAppBar.kt:60]: Yes you're right, this makes way more sense

**Comment by agonzalez-r** [comment]
> I realize there are quite a few changes in the tests, mainly because I didn't test thoroughly while implementing navigation. Sorry for the extra work!

no worries, they are minor changes but they do make the code a lot better

File: app/src/androidTest/java/com/android/periodpals/ui/navigation/TopAppBarTest.kt [hunk 5]
```diff
@@ -29,6 +29,14 @@ class TopAppBarTest {
     composeTestRule.onNodeWithTag("goBackButton").assertIsDisplayed()
   }
 
+  @Test
+  fun topAppBar_displaysEditButton() {
```
Comment (taghizadlaura) [inline app/src/androidTest/java/com/android/periodpals/ui/navigation/TopAppBarTest.kt:33]: The test name doesn't match the naming convention of our other test classes, such as `AlertScreenTest.kt`, which we recently merged. Please try to follow the format used in names like `createInvalidAlertNoProduct`, as this will also help reduce the length of the names, making them more readable.

**Review by taghizadlaura — CHANGES_REQUESTED** [review]
## Summary
Unless what France said and my comment, it LGTM. 
The `TopAppBarTest` class contains unit tests for the `TopAppBar` component, ensuring that it behaves as expected regarding its title, button visibility, and click functionality.

## Important
### Code Quality
- **Readability**: The code is well-structured with clear test names that indicate what each test is verifying. However, the names could be shortened, and fit more the other tests for improved readability.
- **Consistency**: PTAL, The test names do not match the naming conventions used in other classes, such as the `Alert` class (e.g., `createInvalidAlertNoProduct`). Establishing a consistent naming pattern across all test classes would improve maintainability and clarity.

### Functionality
If you make the changes France requested :
- **Correctness**: LGTM, Each test checks specific features of the `TopAppBar`, ensuring that titles and buttons appear correctly based on the provided parameters.
- **Edge Cases**: LGTM, Tests handle scenarios where buttons are not displayed or when click handlers are null, asserting that appropriate exceptions are thrown or behaviors occur.

### Performance
- Nothing to add.

### Security
- Nothing to add.

## Testing
- If you make the changes France requested, it LGTM.

## Documentation
- LGTM, The test methods are self-explanatory (additional comments could enhance clarity about the purpose of specific tests or edge cases but it's not relevant).

## Suggestions
- Make the changes France requested
- Match the naming conventions used in other classes for test names

## Steps before PR approved
1. Change what France requested
2. Try to match the naming convention used in other classes for test names

File: app/src/androidTest/java/com/android/periodpals/ui/profile/CreateProfileTest.kt
Comment (francelu) [inline app/src/androidTest/java/com/android/periodpals/ui/profile/CreateProfileTest.kt:1]: LGTM. I think this is from a merge with main into the branch.

File: app/src/main/java/com/android/periodpals/ui/profile/CreateProfile.kt
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/profile/CreateProfile.kt:1]: LGTM. Seems like a merge with main into the branch too.

**Review by francelu — APPROVED** [review]
LGTM. Well done ! :))

**Review by taghizadlaura — APPROVED** [review]
now it LGTM :)
